Adds structured-RerF parameters to fastRerF

MrAE opened the pull request.

MrAE committed:
Add Structured-RerF to `binnnedBaseTern` as methodToUse=2.

add methodToUse.

add default to switch statement.

MrAE committed:
add structured-RerF test.

MrAE committed:
Merge branch 'staging' into jlp/s-rerf

MrAE committed:
add structured-rerf parameters to fastRerF.

MrAE committed:
add tests for S-RerF

MrAE committed:
update test for s-rerf

MrAE committed:
Merge branch 'staging' into jlp/s-rerf

MrAE committed:
Merge branch 'staging' into jlp/s-rerf_python

MrAE committed:
Merge branch 'jlp/s-rerf' into jlp/s-rerf_python

MrAE committed:
add throw statement when `methodToUse` is set

MrAE committed:
remove default in switch statement

MrAE committed:
add test for paramsRandMatImagePatch
convert `heightWidthTop` to vector of vectors for ease of interpretability.

MrAE committed:
Merge branch 'staging' into jlp/s-rerf

MrAE committed:
Merge branch 'jlp/s-rerf' into jlp/s-rerf_python

falkben commented on Python/examples/example.py:
```diff
@@ -38,10 +38,18 @@
 #     trees=500,
 #     numCores=cpu_count(),
 # )
+#
```
instead of modifying this example.py (which isn't very easy to discover) -- I think we should create a new notebook showing how s-rerf does vs. sklearn/ternary/binary for mnist possibly. which can go on the docs page. :)

falkben replied:
one more thing -- can you add a quick test to show that s-rerf is actually callable from Python?

just a quick addition to `test_RerF.py`

MrAE replied:
Done. https://github.com/neurodata/RerF/blob/c6febfee1dd10e4e8e94698e3b2a21e3290446b7/Python/tests/test_RerF.py#L169

falkben reviewed (changes requested):
generally same comments as R code (make it more user friendly by avoiding the user needing to specify `methodToUse`

falkben committed:
Merge branch 'staging' into jlp/s-rerf_python

falkben committed:
Ben/s rerf sklearn style (#274)

* add s-rerf to rerfClassifier (using methodToUse)

adds test for s-rerfClassifier

adds two jupyter notebook demos for s-rerf (need tuning)

removed imports from sklearn_classifier notebook

* added the notebooks to the docs

minor updates to the notebooks (appearance)

MrAE committed:
Merge branch 'staging' into jlp/s-rerf_python

MrAE committed:
JLP: add tests (#281)

* add check for filling out the grid.

* add test and fix forestType instantiation.

* add comment

falkben committed:
Merge branch 'staging' into jlp/s-rerf_python

falkben commented on Python/rerf/RerF.py:
```diff
@@ -17,6 +17,12 @@ def fastRerF(
+    imageWidth=0,
+    patchHeightMax=0,
+    patchHeightMin=0,
+    patchWidthMax=0,
```
should these default to 0?

MrAE replied:
We have no way of knowing what they should be and 0 will yield a runtime error with a message.

falkben replied:
patch height/widths could have intelligent defaults...  also fine with leaving this as is since we're likely going to deprecate this submodule

falkben commented on Python/rerf/RerF.py:
```diff
@@ -79,7 +85,21 @@ def fastRerF(
+        forestClass.setParameter("imageWidth", imageWidth)
+        forestClass.setParameter("patchHeightMax", patchHeightMax)
+        forestClass.setParameter("patchHeightMin", patchHeightMin)
+        forestClass.setParameter("patchWidthMax", patchWidthMax)
```
add checking here to make sure min isn't > max and min is > 1 and max <= height/width

MrAE replied:
We check those things in C++ ...

falkben commented on Python/rerf/rerfClassifier.py:
```diff
@@ -73,6 +76,21 @@ class rerfClassifier(BaseEstimator, ClassifierMixin):
+        If None, set to ``image_height``.
+    patch_height_min : int, optional (default=1)
+        S-RerF parameter. Minimum image patch height to randomly select from.
+    patch_width_max : int, optional (default=None)
```
i was thinking about this. maybe default should be sqrt of image height/width for image width/height instead of max height/width?

MrAE replied:
I like that much better than the full extent.

MrAE replied:
Done

falkben committed:
update nb w/ very basic comparison to RerF/s-rerf

falkben commented on Python/tests/test_RerF.py:
```diff
@@ -164,3 +164,69 @@ def test_fastPredictPost_X_Y_input():
+    feat_data = train_X[:, 1:]  # mnist
+    labels = train_X[:, label_col]
+
+    ## Subset the data to for time.
```
i think i know what you mean here, but specifically, it does not make sense

MrAE replied:
fixed, thanks.

falkben commented on Python/tests/test_rerfClassifier.py:
```diff
@@ -77,6 +78,71 @@ def test_attributes():
+    assert clf.image_width == 8
+    assert clf.patch_width_max_ == 8
+    assert clf.patch_width_min == 1
+    assert clf.patch_height_max_ == 8
```
some of these have underscores and some do not...  is there a reason for that?

MrAE replied:
Fixed.

MrAE committed:
Merge branch 'staging' into jlp/s-rerf_python

falkben commented on packedForest/test/uniformStructureRerFTest.R:
```diff
@@ -0,0 +1,20 @@
+
+
+h <- read.table('heights.txt')[[1]]
```
not sure these txt file are even in the repo...?

MrAE replied:
See https://github.com/neurodata/RerF/blob/jlp/s-rerf_python/packedForest/test/fpTests/binnedForest/processingNodeBinTest.h#L236

MrAE replied:
We hadn't quite figured out how to run the appropriate test.  There is code that generates those files that is commented out, and the R-script is there to run manually.

MrAE committed:
Make setting parameters fo S-RerF more user-friendly.

MrAE committed:
Merge branch 'staging' into jlp/s-rerf_python

MrAE committed:
undo changes to examples.py

MrAE committed:
update comments in tests.

falkben commented on Python/rerf/rerfClassifier.py:
```diff
@@ -162,17 +196,59 @@ def fit(self, X, y):
+                self.image_width_ = self.image_width
+            self.forest_.setParameter("imageWidth", self.image_width_)
+            # If patch_height_{min, max} and patch_width_{min, max} are
+            # not set by teh user, set them to defaults.
```
teh -> the

falkben commented on Python/tests/test_RerF.py:
```diff
@@ -164,3 +164,72 @@ def test_fastPredictPost_X_Y_input():
+        X=feat_data,
+        Y=labels,
+        forestType="S-RerF",
+        trees=500,
```
in test, we can do fewer trees?

MrAE replied:
Of course, changed to 10.

falkben commented on Python/tests/test_RerF.py:
```diff
@@ -164,3 +164,72 @@ def test_fastPredictPost_X_Y_input():
+    # Map output labels (0,1) --> (3,5)
+    train_pred = [[3,5][i] for i in fastPredict(feat_data, forest)]
+
+    print("Trianing Error:", np.mean(train_pred != train_X[y35, label_col]))
```
generally want to avoid print statements in tests

falkben reviewed (changes requested):
very minor changes requested. nice!

MrAE committed:
updates from Ben's comments

MrAE committed:
Merge branch 'staging' into jlp/s-rerf_python

falkben commented on Python/rerf/rerfClassifier.py:
```diff
@@ -168,23 +202,62 @@ def fit(self, X, y):
         # setup the forest's parameters
         self.forest_ = pyfp.fpForest()
 
-        if self.projection_matrix == "RerF":
```
well.  here it is. reason why OOB stuff isn't working.  your s-rerf PR conflicted. and neither of us caught it.  and i didn't have any tests for oob so travis didn't fail.
